File: source/go/guitar.go
// imports
package main
import (
    "fmt"
    "os"
    "bufio"
    "strings"
)

// main
func main() {
    discourse := []string{"a", "A", "b", "c", "C", "d", "D", "e", "f", "F", "g", "G"}
    g := guitar{discourse}

    for {
        input  := menu(">")
        action := strings.Split(input, " ")
        g.dispatch(action[0], action[1:])
    }
}

// functions
func rotate(list []string, by int) []string {
    for i := 0; i < by; i++ {
        head := list[0]
        tail := list[1:]
        list  = append(tail, head)
    }
    return list
}
func fret(note string) string {
    return "[" + note + "]-"
}
func line(list []string, by int) string {
    rotated := rotate(list, by)

    var line string
    for _, note := range rotated {
        line = line + fret(note)
    }
    return line
}
func isolate(list []string, filters []string) []string {
    var filtered []string
    element := "_"

    for _, note := range list {
        for _, filter := range filters {
            if note == filter {
                element = note
                break
            }
        }
        filtered = append(filtered, element)
        element = "_"
    }
    return filtered
}
func fretboard(discourse []string, filters []string) {
    fmt.Println(" 0           3       5       7       9")
    fmt.Println(line(isolate(discourse, filters),  7))
    fmt.Println(line(isolate(discourse, filters),  2))
    fmt.Println(line(isolate(discourse, filters), 10))
    fmt.Println(line(isolate(discourse, filters),  5))
    fmt.Println(line(isolate(discourse, filters),  0))
    fmt.Println(line(isolate(discourse, filters),  7))
    fmt.Println(" 0           3       5       7       9")
}
func menu(prompt string) string {
    reader := bufio.NewReader(os.Stdin)
    fmt.Print(prompt + " ")
    text, _ := reader.ReadString('\n')
    return strings.Trim(text, "\n ")
}

// guitar
type guitar struct {
    notes []string
}
func (g *guitar) dispatch(command string, arguments []string) {
    switch command {
    case "fret":
        fretboard(g.notes, g.notes)
    case "scale":
        fretboard(g.notes, arguments)
    case "exit":
        os.Exit(0)
    }
}

File: source/swift/guitar.swift
// imports
import Foundation

// functions
func rotate(list: [String], by: Int) -> [String] {
    var rotated = list
    for _ in 0..<by {
        let head = rotated[0]
        let tail = rotated[1..<list.endIndex]
        rotated = tail + [head]
    }
    return rotated
}
func fret(note: String) -> String {
    return "[" + note + "]-"
}
func line(list: [String], by: Int) -> String {
    let rotated = rotate(list: list, by: by)
    var line = String()

    for note in rotated {
        line += fret(note: note)
    }
    return line
}
func isolate(list: [String], choices: [String]) -> [String] {
    return list.map({
        (e: String) -> String in
        return choices.contains(e) ? e : "_"
    })
}
func fretboard(discourse: [String], choices: [String]) {
    print(" 0           3       5       7       9");
    print(line(list: isolate(list: discourse, choices: choices), by:  7))
    print(line(list: isolate(list: discourse, choices: choices), by:  2))
    print(line(list: isolate(list: discourse, choices: choices), by: 10))
    print(line(list: isolate(list: discourse, choices: choices), by:  5))
    print(line(list: isolate(list: discourse, choices: choices), by:  0))
    print(line(list: isolate(list: discourse, choices: choices), by:  7))
    print(" 0           3       5       7       9");
}
func menu(prompt: String) -> String {
    print("\(prompt) ", terminator:"")
    let input = readLine(strippingNewline: true)
    return input ?? ""
}
func parse(input: String) -> [String] {
    let trim = input.trimmingCharacters(in: .whitespacesAndNewlines)
    return trim.components(separatedBy: " ")
}
func start() {
    let discourse = ["a", "A", "b", "c", "C", "d", "D", "e", "f", "F", "g", "G"]
    
    while true {
        let input     = menu(prompt: ">")
        let action    = parse(input: input)
        let command   = action[0]
        let arguments = Array(action[1..<action.endIndex])

        if command == "fret" {
            fretboard(discourse: discourse, choices: discourse)
        } else if command == "scale" {
            fretboard(discourse: discourse, choices: arguments)
        } else if command == "exit" {
            exit(0)
        }
    }
}

// main
func main() {
    start()
}
main()
